File: rasterTest/src/main/java/com/kjipo/raster/FlowDirection.java
package com.kjipo.raster;


public enum FlowDirection {
    EAST(0, 1, 0),
    NORTH_EAST(-1, 1, Math.PI / 4),
    NORTH(-1, 0, Math.PI / 2),
    NORTH_WEST(-1, -1, 3 * Math.PI / 4),
    WEST(0, -1, Math.PI),
    SOUTH_WEST(1, -1, 5 * Math.PI / 4),
    SOUTH(1, 0, 3 * Math.PI / 2),
    SOUTH_EAST(1, 1, 7 * Math.PI / 4);

    private final int rowShift;
    private final int columnShift;
    private final double angleInRadians;


    FlowDirection(int rowShift, int columnShift, double angleInRadians) {
        this.rowShift = rowShift;
        this.columnShift = columnShift;
        this.angleInRadians = angleInRadians;
    }

    public int getRowShift() {
        return rowShift;
    }

    public int getColumnShift() {
        return columnShift;
    }

    public double getAngleInRadians() {
        return angleInRadians;
    }

}

File: encodings-multimodule/src/commonMain/kotlin/com/kjipo/representation/raster/FlowDirection.kt
package com.kjipo.representation.raster

import kotlin.math.PI


enum class FlowDirection(val rowShift: Int, val columnShift: Int, val angleInRadians: Double) {
    EAST(0, 1, 0.0),
    NORTH_EAST(-1, 1, PI / 4),
    NORTH(-1, 0, PI / 2),
    NORTH_WEST(-1, -1, 3 * PI / 4),
    WEST(0, -1, PI),
    SOUTH_WEST(1, -1, 5 * PI / 4),
    SOUTH(1, 0, 3 * PI / 2),
    SOUTH_EAST(1, 1, 7 * PI / 4)


}

fun getFlowDirectionForOffset(rowShift: Int, columnShift: Int): FlowDirection? {
   return FlowDirection.values().find { it.rowShift == rowShift && it.columnShift == columnShift }
}

fun FlowDirection.shiftTwoStepsForward(): FlowDirection {
    return FlowDirection.values()[(this.ordinal + 2) % FlowDirection.values().size]
}

